File: custom-exception/src/main/java/dev/trile/customexception/MoneyTransfer.java
package dev.trile.customexception;

import static dev.trile.customexception.ErrorType.TRANSFER_MONEY_NOT_ENOUGH_MONEY;

public class MoneyTransfer {

  private final Bank primaryBank = new VietcomBank();

  private final Bank secondaryBank = new VietinBank();

  public void transferMoney(String accountNo, int amount) throws SalaryException {
    System.out.println("Transfer " + amount + " to " + accountNo);

    try {
      primaryBank.transferMoney(accountNo, amount);
    } catch (SalaryException e) {
      System.out.println(e.getMessage());

      if (e.getCodeType() != TRANSFER_MONEY_NOT_ENOUGH_MONEY)
        throw e; // rethrow exception

      System.out.println("Retry with VTB");
      // try with VietinBank
      secondaryBank.transferMoney(accountNo, amount);
    }
  }
}

File: ioc-di/src/main/kotlin/dev/trile/iocdi/MoneyTransfer.kt
package dev.trile.iocdi

import org.springframework.stereotype.Component

@Component
class MoneyTransfer(
  private val primaryBank: Bank,
  private val secondaryBank: Bank
) {

  fun transferMoney(accountNo: String, amount: Int) {
    println("Transfer $amount to $accountNo")
    try {
      primaryBank.transferMoney(accountNo, amount)
    } catch (e: SalaryException) {
      println(e.message)
      if (e.codeType !== ErrorType.TRANSFER_MONEY_NOT_ENOUGH_MONEY) throw e // rethrow exception
      println("Retry with VTB")
      // try with VietinBank
      secondaryBank.transferMoney(accountNo, amount)
    }
  }
}
